core/byzantine_test.go

```diff
@@ -0,0 +1,587 @@
+package core
+
+import (
+	"bytes"
+	"github.com/0xPolygon/go-ibft/messages/proto"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestByzantineBehaviour(t *testing.T) {
+	t.Parallel()
+
+	t.Run("malicious hash in proposal", func(t *testing.T) {
+		t.Parallel()
+
+		cluster := newCluster(
+			6,
+			func(c *cluster) {
+				for _, node := range c.nodes {
+					currentNode := node
+					node.core = NewIBFT(
+						mockLogger{},
+						&mockBackend{
+							isValidBlockFn:         isValidProposal,
+							isValidProposalHashFn:  isValidProposalHash,
+							isValidSenderFn:        nil,
+							isValidCommittedSealFn: nil,
+							isProposerFn: func(from []byte, height uint64, round uint64) bool {
+								if round == 0 {
+									return false
+								}
+
+								return bytes.Equal(
+									from,
+									c.addresses()[int(round)%len(c.addresses())],
+								)
+							},
+
+							idFn: node.addr,
+
+							buildProposalFn: buildValidProposal,
+							buildPrePrepareMessageFn: func(proposal []byte,
+								rcc *proto.RoundChangeCertificate,
+								view *proto.View) *proto.Message {
+								proposalHash := validProposalHash
+								if currentNode.byzantine {
+									proposalHash = []byte("invalid proposal hash")
+								}
+
+								return buildBasicPreprepareMessage(
+									proposal,
+									proposalHash,
+									rcc,
+									currentNode.address,
+									view,
+								)
+							},
+							buildPrepareMessageFn:     node.buildPrepare,
+							buildCommitMessageFn:      node.buildCommit,
+							buildRoundChangeMessageFn: node.buildRoundChange,
+
+							insertBlockFn: nil,
+							hasQuorumFn:   c.hasQuorumFn,
+						},
```

Comment:
Most fields in backend are common, shall we create a helper function to create backend with common fields for shorter test like?

```go
func TestByzantineBehaviour(t *testing.T) {
	t.Parallel()

	newBaseBackend := func(c *cluster, node *node) *mockBackend {
		return &mockBackend{
			isValidBlockFn:         isValidProposal,
			isValidProposalHashFn:  isValidProposalHash,
			isValidSenderFn:        nil,
			isValidCommittedSealFn: nil,
			isProposerFn:           c.isProposer,
                    ....
		}
	}

...

	t.Run("malicious hash in proposal", func(t *testing.T) {
		t.Parallel()

		cluster := newCluster(
			6,
			func(c *cluster) {
				for _, node := range c.nodes {
					currentNode := node

					backend := newBaseBackend(c, node)

					backend.isProposerFn = func(from []byte, height uint64, round uint64) bool {
						if round == 0 {
							return false
						}

						return bytes.Equal(
							from,
							c.addresses()[int(round)%len(c.addresses())],
						)
					}
```